packages/@glimmer/validator/lib/tracking.ts

```diff
@@ -96,64 +115,122 @@ export function memo<T, U, V, W, X, Y, Z>(
 ): (a1: U, a2: V, a3: W, a4: X, a5: Y, a6: Z) => T;
 
 export function memo<T>(callback: (...args: unknown[]) => T, debuggingContext?: string | false) {
-  let lastValue: T | undefined;
-  let tag: Tag;
-  let snapshot: Revision;
+  let cache = new Cache(callback, debuggingContext);
+
+  let memoized = (...args: unknown[]) => getValue(cache, ...args);
+
+  (memoized as any)[CACHE_KEY] = cache;
+
+  return memoized;
+}
+
+export function isConstMemo(memoized: Function) {
+  let cache = (memoized as any)[CACHE_KEY] as Cache | undefined;
+
+  return cache === undefined ? false : isConst(cache);
+}
+
+//////////
+
+let DEBUG_CACHE = symbol('DEBUG_CACHE');
+let DEBUG_CACHE_CONTEXT = symbol('DEBUG_CACHE_CONTEXT');
+
+class Cache<T = unknown> {
+  private lastValue: T | undefined;
+  private tag: Tag | undefined;
+  private snapshot: Revision = -1;
+
+  constructor(private fn: (...args: unknown[]) => T, debuggingContext?: string | false) {
+    if (DEBUG) {
+      return ({
+        [DEBUG_CACHE]: this,
+        [DEBUG_CACHE_CONTEXT]: debuggingContext,
+      } as any) as Cache<T>;
```

Comment:
Seems somewhat bad to return a different thing than `this` in a constructor. I know it works, and why, but seems like a pretty big footgun and I'm not really sure what all this complexity buys us.